{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aim: Integrate gene expression data from three online data portals: Stemformatics, ENCODE and Haemosphere. \n",
    "\n",
    "In previous notebooks, we have combined the gene expression data of selected samples in each individual data portals. Here, we will integrate the three pre-combined expression datasets together into a larger dataset of mouse blood. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/monica/anaconda3/envs/stemformatics/lib/python3.7/site-packages/anndata/core/anndata.py:17: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import RangeIndex\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import atlas\n",
    "import handler\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge expression tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data \n",
    "expression_s4m    = pd.read_csv('../data/interim/mouse_integrate/expression_s4m.tsv', sep='\\t', index_col=0)\n",
    "expression_encode = pd.read_csv('../data/interim/mouse_integrate/expression_encode.tsv', sep='\\t', index_col=0)\n",
    "expression_haemosphere = pd.read_csv('../data/interim/mouse_integrate/expression_Haemosphere.tsv', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18483, 217) (22457, 90) (14462, 786)\n"
     ]
    }
   ],
   "source": [
    "print(expression_s4m.shape, expression_encode.shape, expression_haemosphere.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14222, 1093)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENCFF547CXK</th>\n",
       "      <th>ENCFF774DIF</th>\n",
       "      <th>ENCFF155RUP</th>\n",
       "      <th>ENCFF702TOW</th>\n",
       "      <th>ENCFF963EBQ</th>\n",
       "      <th>ENCFF649QOI</th>\n",
       "      <th>ENCFF871RIM</th>\n",
       "      <th>ENCFF415ZGH</th>\n",
       "      <th>ENCFF924EMV</th>\n",
       "      <th>ENCFF063SJF</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM854333</th>\n",
       "      <th>GSM854334</th>\n",
       "      <th>GSM1023629</th>\n",
       "      <th>GSM1023630</th>\n",
       "      <th>GSM1023631</th>\n",
       "      <th>GSM1023632</th>\n",
       "      <th>GSM1023633</th>\n",
       "      <th>GSM1023634</th>\n",
       "      <th>GSM1023635</th>\n",
       "      <th>GSM1023636</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.941569</td>\n",
       "      <td>0.931585</td>\n",
       "      <td>0.711538</td>\n",
       "      <td>0.716847</td>\n",
       "      <td>0.949093</td>\n",
       "      <td>0.945296</td>\n",
       "      <td>0.901315</td>\n",
       "      <td>0.923745</td>\n",
       "      <td>0.844572</td>\n",
       "      <td>0.793735</td>\n",
       "      <td>...</td>\n",
       "      <td>0.872521</td>\n",
       "      <td>0.863803</td>\n",
       "      <td>0.949937</td>\n",
       "      <td>0.953312</td>\n",
       "      <td>0.943257</td>\n",
       "      <td>0.943679</td>\n",
       "      <td>0.939812</td>\n",
       "      <td>0.944101</td>\n",
       "      <td>0.959007</td>\n",
       "      <td>0.956757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.135951</td>\n",
       "      <td>0.135319</td>\n",
       "      <td>0.097384</td>\n",
       "      <td>0.109127</td>\n",
       "      <td>0.129764</td>\n",
       "      <td>0.129623</td>\n",
       "      <td>0.169702</td>\n",
       "      <td>0.182956</td>\n",
       "      <td>0.183835</td>\n",
       "      <td>0.179862</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001195</td>\n",
       "      <td>0.001617</td>\n",
       "      <td>0.005484</td>\n",
       "      <td>0.002391</td>\n",
       "      <td>0.002531</td>\n",
       "      <td>0.005836</td>\n",
       "      <td>0.005414</td>\n",
       "      <td>0.004781</td>\n",
       "      <td>0.002602</td>\n",
       "      <td>0.002883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.470855</td>\n",
       "      <td>0.465265</td>\n",
       "      <td>0.746941</td>\n",
       "      <td>0.785297</td>\n",
       "      <td>0.854627</td>\n",
       "      <td>0.863135</td>\n",
       "      <td>0.812720</td>\n",
       "      <td>0.804212</td>\n",
       "      <td>0.883315</td>\n",
       "      <td>0.910561</td>\n",
       "      <td>...</td>\n",
       "      <td>0.764942</td>\n",
       "      <td>0.774012</td>\n",
       "      <td>0.784419</td>\n",
       "      <td>0.854099</td>\n",
       "      <td>0.814231</td>\n",
       "      <td>0.901209</td>\n",
       "      <td>0.893123</td>\n",
       "      <td>0.882365</td>\n",
       "      <td>0.866053</td>\n",
       "      <td>0.709605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.135951</td>\n",
       "      <td>0.135319</td>\n",
       "      <td>0.428069</td>\n",
       "      <td>0.446105</td>\n",
       "      <td>0.414674</td>\n",
       "      <td>0.426136</td>\n",
       "      <td>0.432675</td>\n",
       "      <td>0.747187</td>\n",
       "      <td>0.553860</td>\n",
       "      <td>0.429300</td>\n",
       "      <td>...</td>\n",
       "      <td>0.031852</td>\n",
       "      <td>0.060470</td>\n",
       "      <td>0.321579</td>\n",
       "      <td>0.478976</td>\n",
       "      <td>0.387920</td>\n",
       "      <td>0.318802</td>\n",
       "      <td>0.405850</td>\n",
       "      <td>0.348123</td>\n",
       "      <td>0.210800</td>\n",
       "      <td>0.361482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>0.135951</td>\n",
       "      <td>0.135319</td>\n",
       "      <td>0.097384</td>\n",
       "      <td>0.109127</td>\n",
       "      <td>0.129764</td>\n",
       "      <td>0.129623</td>\n",
       "      <td>0.169702</td>\n",
       "      <td>0.182956</td>\n",
       "      <td>0.183835</td>\n",
       "      <td>0.179862</td>\n",
       "      <td>...</td>\n",
       "      <td>0.072775</td>\n",
       "      <td>0.056673</td>\n",
       "      <td>0.091759</td>\n",
       "      <td>0.074251</td>\n",
       "      <td>0.098650</td>\n",
       "      <td>0.070876</td>\n",
       "      <td>0.075657</td>\n",
       "      <td>0.077978</td>\n",
       "      <td>0.149557</td>\n",
       "      <td>0.110533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1093 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENCFF547CXK  ENCFF774DIF  ENCFF155RUP  ENCFF702TOW  \\\n",
       "ENSMUSG00000000001     0.941569     0.931585     0.711538     0.716847   \n",
       "ENSMUSG00000000003     0.135951     0.135319     0.097384     0.109127   \n",
       "ENSMUSG00000000028     0.470855     0.465265     0.746941     0.785297   \n",
       "ENSMUSG00000000037     0.135951     0.135319     0.428069     0.446105   \n",
       "ENSMUSG00000000049     0.135951     0.135319     0.097384     0.109127   \n",
       "\n",
       "                    ENCFF963EBQ  ENCFF649QOI  ENCFF871RIM  ENCFF415ZGH  \\\n",
       "ENSMUSG00000000001     0.949093     0.945296     0.901315     0.923745   \n",
       "ENSMUSG00000000003     0.129764     0.129623     0.169702     0.182956   \n",
       "ENSMUSG00000000028     0.854627     0.863135     0.812720     0.804212   \n",
       "ENSMUSG00000000037     0.414674     0.426136     0.432675     0.747187   \n",
       "ENSMUSG00000000049     0.129764     0.129623     0.169702     0.182956   \n",
       "\n",
       "                    ENCFF924EMV  ENCFF063SJF  ...  GSM854333  GSM854334  \\\n",
       "ENSMUSG00000000001     0.844572     0.793735  ...   0.872521   0.863803   \n",
       "ENSMUSG00000000003     0.183835     0.179862  ...   0.001195   0.001617   \n",
       "ENSMUSG00000000028     0.883315     0.910561  ...   0.764942   0.774012   \n",
       "ENSMUSG00000000037     0.553860     0.429300  ...   0.031852   0.060470   \n",
       "ENSMUSG00000000049     0.183835     0.179862  ...   0.072775   0.056673   \n",
       "\n",
       "                    GSM1023629  GSM1023630  GSM1023631  GSM1023632  \\\n",
       "ENSMUSG00000000001    0.949937    0.953312    0.943257    0.943679   \n",
       "ENSMUSG00000000003    0.005484    0.002391    0.002531    0.005836   \n",
       "ENSMUSG00000000028    0.784419    0.854099    0.814231    0.901209   \n",
       "ENSMUSG00000000037    0.321579    0.478976    0.387920    0.318802   \n",
       "ENSMUSG00000000049    0.091759    0.074251    0.098650    0.070876   \n",
       "\n",
       "                    GSM1023633  GSM1023634  GSM1023635  GSM1023636  \n",
       "ENSMUSG00000000001    0.939812    0.944101    0.959007    0.956757  \n",
       "ENSMUSG00000000003    0.005414    0.004781    0.002602    0.002883  \n",
       "ENSMUSG00000000028    0.893123    0.882365    0.866053    0.709605  \n",
       "ENSMUSG00000000037    0.405850    0.348123    0.210800    0.361482  \n",
       "ENSMUSG00000000049    0.075657    0.077978    0.149557    0.110533  \n",
       "\n",
       "[5 rows x 1093 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = [expression_encode, expression_haemosphere, expression_s4m]\n",
    "common_genes = handler.find_common_genes(dfs)\n",
    "mouse_atlas_expression = atlas.rankTransform(handler.merge_columns(dfs, common_genes))\n",
    "\n",
    "print(mouse_atlas_expression.shape)\n",
    "mouse_atlas_expression.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Standardise sample metadata \n",
    "\n",
    "We have encountered several difficulties during the integration of the metadata that associated with the collected datasets. \n",
    "\n",
    "1. Different datasets may record different pieces of information to describe each sample. e.g. 'tissue' attribute is recorded in some sample metadata but not all.\n",
    "2. Same piece of information can be recorded in different ways. e.g. cell type information is recorded in a single 'celltype' column in the haemosphere metadata, whereas in s4m metadata, there are 4 columns having information related to the cell type of samples.\n",
    "3. Inconsistent format of contents e.g. macrophage is represented as 'BM mac', 'BM macrophage', 'BM-derived macrophage day 0', 'Bone marrow derived macrophage' in a same column. \n",
    "4. Different type of information might be stored mixedly under the same attributes. e.g. In 'replicate_group_id' column of the s4m metadata, we might find experiment information, cell type and sort markers information about samples.\n",
    "\n",
    "To address these issues:\n",
    "1. we will determin a essential list of attributes to describe samples and unify the naming of these attributes. These essential list of attributes are: \n",
    "\n",
    "       Cell Type; Cell Lineage; Description; Dataset Name; Platform \n",
    "\n",
    "2. Reannotate the metadata so that the content of each attribute is consistent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the reannotated sample metadata \n",
    "# For metadata of each data collection: stemformatics, encode and haemosphere, we added two manually annotated columns\n",
    "# 'cell_lineage_anno' and 'celltype_anno' with consistent content format. \n",
    "samples_s4m = pd.read_csv('../data/interim/reannotated/samples_s4m_anno.tsv', sep='\\t', index_col=0)\n",
    "samples_encode = pd.read_csv('../data/interim/reannotated/samples_encode_anno.tsv', sep='\\t', index_col=0)\n",
    "samples_haemosphere = pd.read_csv('../data/interim/reannotated/samples_Haemosphere_anno.tsv', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "slice relevant columns from each metadata table and rename in consistent format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ds_id</th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>replicate_group_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>generic_sample_type</th>\n",
       "      <th>final_cell_type</th>\n",
       "      <th>parental_cell_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>labelling</th>\n",
       "      <th>description</th>\n",
       "      <th>platform</th>\n",
       "      <th>cell_lineage_anno</th>\n",
       "      <th>celltype_anno</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chip_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM98876</th>\n",
       "      <td>6658.0</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Loss of immune function and an increased incid...</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>HSC_HPC</td>\n",
       "      <td>hematopoietic stem cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM98877</th>\n",
       "      <td>6658.0</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Loss of immune function and an increased incid...</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>HSC_HPC</td>\n",
       "      <td>hematopoietic stem cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM98878</th>\n",
       "      <td>6658.0</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Loss of immune function and an increased incid...</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>HSC_HPC</td>\n",
       "      <td>hematopoietic stem cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ds_id dataset_name  \\\n",
       "chip_id                         \n",
       "GSM98876  6658.0        Rossi   \n",
       "GSM98877  6658.0        Rossi   \n",
       "GSM98878  6658.0        Rossi   \n",
       "\n",
       "                                         replicate_group_id      organism  \\\n",
       "chip_id                                                                     \n",
       "GSM98876  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "GSM98877  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "GSM98878  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "\n",
       "                 sample_type generic_sample_type final_cell_type  \\\n",
       "chip_id                                                            \n",
       "GSM98876  LT-HSC Aged animal  LT-HSC Aged animal             NaN   \n",
       "GSM98877  LT-HSC Aged animal  LT-HSC Aged animal             NaN   \n",
       "GSM98878  LT-HSC Aged animal  LT-HSC Aged animal             NaN   \n",
       "\n",
       "         parental_cell_type  sex labelling  \\\n",
       "chip_id                                      \n",
       "GSM98876                NaN  NaN       NaN   \n",
       "GSM98877                NaN  NaN       NaN   \n",
       "GSM98878                NaN  NaN       NaN   \n",
       "\n",
       "                                                description  \\\n",
       "chip_id                                                       \n",
       "GSM98876  Loss of immune function and an increased incid...   \n",
       "GSM98877  Loss of immune function and an increased incid...   \n",
       "GSM98878  Loss of immune function and an increased incid...   \n",
       "\n",
       "                                    platform cell_lineage_anno  \\\n",
       "chip_id                                                          \n",
       "GSM98876  Microarray (Affymetrix Mouse430_2)           HSC_HPC   \n",
       "GSM98877  Microarray (Affymetrix Mouse430_2)           HSC_HPC   \n",
       "GSM98878  Microarray (Affymetrix Mouse430_2)           HSC_HPC   \n",
       "\n",
       "                    celltype_anno  \n",
       "chip_id                            \n",
       "GSM98876  hematopoietic stem cell  \n",
       "GSM98877  hematopoietic stem cell  \n",
       "GSM98878  hematopoietic stem cell  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_s4m[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_s4m = samples_s4m[['celltype_anno', 'cell_lineage_anno', 'description', 'platform', 'dataset_name']]\n",
    "samples_s4m.columns = ['Cell Type', 'Cell Lineage', 'Description', 'Platform', 'Dataset Name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>description</th>\n",
       "      <th>markers</th>\n",
       "      <th>platform</th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>cell_lineage_anno</th>\n",
       "      <th>celltype_anno</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561641</th>\n",
       "      <td>Fob</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Follicular B cells</td>\n",
       "      <td>FSC-lo B220:+ CD21:+ CD23:+</td>\n",
       "      <td>RNAseq (Illumina HiSeq)</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "      <td>B-cell</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561642</th>\n",
       "      <td>Fob</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Follicular B cells</td>\n",
       "      <td>FSC-lo B220:+ CD21:+ CD23:+</td>\n",
       "      <td>RNAseq (Illumina HiSeq)</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "      <td>B-cell</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561645</th>\n",
       "      <td>MZB</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Marginal zone B cells</td>\n",
       "      <td>FSC-lo B220:+ CD21:-hi CD23:-</td>\n",
       "      <td>RNAseq (Illumina HiSeq)</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "      <td>B-cell</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    celltype    cell_lineage            description  \\\n",
       "GSE60927.SRR1561641      Fob  B Cell Lineage     Follicular B cells   \n",
       "GSE60927.SRR1561642      Fob  B Cell Lineage     Follicular B cells   \n",
       "GSE60927.SRR1561645      MZB  B Cell Lineage  Marginal zone B cells   \n",
       "\n",
       "                                           markers                 platform  \\\n",
       "GSE60927.SRR1561641    FSC-lo B220:+ CD21:+ CD23:+  RNAseq (Illumina HiSeq)   \n",
       "GSE60927.SRR1561642    FSC-lo B220:+ CD21:+ CD23:+  RNAseq (Illumina HiSeq)   \n",
       "GSE60927.SRR1561645  FSC-lo B220:+ CD21:-hi CD23:-  RNAseq (Illumina HiSeq)   \n",
       "\n",
       "                                dataset_name cell_lineage_anno celltype_anno  \n",
       "GSE60927.SRR1561641  Haemopedia-Mouse-RNAseq            B-cell        B cell  \n",
       "GSE60927.SRR1561642  Haemopedia-Mouse-RNAseq            B-cell        B cell  \n",
       "GSE60927.SRR1561645  Haemopedia-Mouse-RNAseq            B-cell        B cell  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_haemosphere[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_haemosphere = samples_haemosphere[['celltype_anno', 'cell_lineage_anno', 'description', 'platform', 'dataset_name']]\n",
    "samples_haemosphere.columns = ['Cell Type', 'Cell Lineage', 'Description', 'Platform', 'Dataset Name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay name</th>\n",
       "      <th>Biosample summary</th>\n",
       "      <th>Biosample term name</th>\n",
       "      <th>Description</th>\n",
       "      <th>Lab</th>\n",
       "      <th>Project</th>\n",
       "      <th>Biosample accession</th>\n",
       "      <th>Biological replicate</th>\n",
       "      <th>Technical replicate</th>\n",
       "      <th>Platform</th>\n",
       "      <th>cell_lineage_anno</th>\n",
       "      <th>celltype_anno</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Files</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCFF547CXK</th>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>C57BL/6 neutrophil adult (5-6 weeks)</td>\n",
       "      <td>neutrophil</td>\n",
       "      <td>PSU mouse neutrophil total RNA scriptseqv2 RNA...</td>\n",
       "      <td>Ross Hardison, PennState</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>ENCBS562NSP,ENCBS622TZA</td>\n",
       "      <td>1,2</td>\n",
       "      <td>1</td>\n",
       "      <td>RNAseq (Illumina NextSeq)</td>\n",
       "      <td>granulocyte</td>\n",
       "      <td>neutrophil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF774DIF</th>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>C57BL/6 neutrophil adult (5-6 weeks)</td>\n",
       "      <td>neutrophil</td>\n",
       "      <td>PSU mouse neutrophil total RNA scriptseqv2 RNA...</td>\n",
       "      <td>Ross Hardison, PennState</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>ENCBS562NSP,ENCBS622TZA</td>\n",
       "      <td>1,2</td>\n",
       "      <td>1</td>\n",
       "      <td>RNAseq (Illumina NextSeq)</td>\n",
       "      <td>granulocyte</td>\n",
       "      <td>neutrophil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF155RUP</th>\n",
       "      <td>polyA plus RNA-seq</td>\n",
       "      <td>C57BL/6 megakaryocyte-erythroid progenitor cel...</td>\n",
       "      <td>megakaryocyte-erythroid progenitor cell</td>\n",
       "      <td>RNA-seq on mouse megakaryocyte-erythroid proge...</td>\n",
       "      <td>Ross Hardison, PennState</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>ENCBS190OUH,ENCBS176ENC</td>\n",
       "      <td>2,1</td>\n",
       "      <td>1</td>\n",
       "      <td>RNAseq (Illumina HiSeq)</td>\n",
       "      <td>HSC_HPC</td>\n",
       "      <td>megakaryocyte erythroid progenitor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Assay name  \\\n",
       "Files                             \n",
       "ENCFF547CXK             RNA-seq   \n",
       "ENCFF774DIF             RNA-seq   \n",
       "ENCFF155RUP  polyA plus RNA-seq   \n",
       "\n",
       "                                             Biosample summary  \\\n",
       "Files                                                            \n",
       "ENCFF547CXK               C57BL/6 neutrophil adult (5-6 weeks)   \n",
       "ENCFF774DIF               C57BL/6 neutrophil adult (5-6 weeks)   \n",
       "ENCFF155RUP  C57BL/6 megakaryocyte-erythroid progenitor cel...   \n",
       "\n",
       "                                 Biosample term name  \\\n",
       "Files                                                  \n",
       "ENCFF547CXK                               neutrophil   \n",
       "ENCFF774DIF                               neutrophil   \n",
       "ENCFF155RUP  megakaryocyte-erythroid progenitor cell   \n",
       "\n",
       "                                                   Description  \\\n",
       "Files                                                            \n",
       "ENCFF547CXK  PSU mouse neutrophil total RNA scriptseqv2 RNA...   \n",
       "ENCFF774DIF  PSU mouse neutrophil total RNA scriptseqv2 RNA...   \n",
       "ENCFF155RUP  RNA-seq on mouse megakaryocyte-erythroid proge...   \n",
       "\n",
       "                                  Lab Project      Biosample accession  \\\n",
       "Files                                                                    \n",
       "ENCFF547CXK  Ross Hardison, PennState  ENCODE  ENCBS562NSP,ENCBS622TZA   \n",
       "ENCFF774DIF  Ross Hardison, PennState  ENCODE  ENCBS562NSP,ENCBS622TZA   \n",
       "ENCFF155RUP  Ross Hardison, PennState  ENCODE  ENCBS190OUH,ENCBS176ENC   \n",
       "\n",
       "            Biological replicate Technical replicate  \\\n",
       "Files                                                  \n",
       "ENCFF547CXK                  1,2                   1   \n",
       "ENCFF774DIF                  1,2                   1   \n",
       "ENCFF155RUP                  2,1                   1   \n",
       "\n",
       "                              Platform cell_lineage_anno  \\\n",
       "Files                                                      \n",
       "ENCFF547CXK  RNAseq (Illumina NextSeq)       granulocyte   \n",
       "ENCFF774DIF  RNAseq (Illumina NextSeq)       granulocyte   \n",
       "ENCFF155RUP    RNAseq (Illumina HiSeq)           HSC_HPC   \n",
       "\n",
       "                                  celltype_anno  \n",
       "Files                                            \n",
       "ENCFF547CXK                          neutrophil  \n",
       "ENCFF774DIF                          neutrophil  \n",
       "ENCFF155RUP  megakaryocyte erythroid progenitor  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_encode[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_encode = samples_encode[['celltype_anno', 'cell_lineage_anno', 'Description', 'Platform', 'Project']]\n",
    "samples_encode.columns = ['Cell Type', 'Cell Lineage', 'Description', 'Platform', 'Dataset Name']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge the three metadata tables "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1102, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell Type</th>\n",
       "      <th>Cell Lineage</th>\n",
       "      <th>Description</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Dataset Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCFF547CXK</th>\n",
       "      <td>neutrophil</td>\n",
       "      <td>granulocyte</td>\n",
       "      <td>PSU mouse neutrophil total RNA scriptseqv2 RNA...</td>\n",
       "      <td>RNAseq (Illumina NextSeq)</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF774DIF</th>\n",
       "      <td>neutrophil</td>\n",
       "      <td>granulocyte</td>\n",
       "      <td>PSU mouse neutrophil total RNA scriptseqv2 RNA...</td>\n",
       "      <td>RNAseq (Illumina NextSeq)</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF155RUP</th>\n",
       "      <td>megakaryocyte erythroid progenitor</td>\n",
       "      <td>HSC_HPC</td>\n",
       "      <td>RNA-seq on mouse megakaryocyte-erythroid proge...</td>\n",
       "      <td>RNAseq (Illumina HiSeq)</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF702TOW</th>\n",
       "      <td>megakaryocyte erythroid progenitor</td>\n",
       "      <td>HSC_HPC</td>\n",
       "      <td>RNA-seq on mouse megakaryocyte-erythroid proge...</td>\n",
       "      <td>RNAseq (Illumina HiSeq)</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF963EBQ</th>\n",
       "      <td>megakaryocyte</td>\n",
       "      <td>megakaryocyte</td>\n",
       "      <td>PSU mouse Megakaryocyte 1 ng RNA-seq</td>\n",
       "      <td>RNAseq (Illumina HiSeq)</td>\n",
       "      <td>ENCODE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Cell Type   Cell Lineage  \\\n",
       "ENCFF547CXK                          neutrophil    granulocyte   \n",
       "ENCFF774DIF                          neutrophil    granulocyte   \n",
       "ENCFF155RUP  megakaryocyte erythroid progenitor        HSC_HPC   \n",
       "ENCFF702TOW  megakaryocyte erythroid progenitor        HSC_HPC   \n",
       "ENCFF963EBQ                       megakaryocyte  megakaryocyte   \n",
       "\n",
       "                                                   Description  \\\n",
       "ENCFF547CXK  PSU mouse neutrophil total RNA scriptseqv2 RNA...   \n",
       "ENCFF774DIF  PSU mouse neutrophil total RNA scriptseqv2 RNA...   \n",
       "ENCFF155RUP  RNA-seq on mouse megakaryocyte-erythroid proge...   \n",
       "ENCFF702TOW  RNA-seq on mouse megakaryocyte-erythroid proge...   \n",
       "ENCFF963EBQ               PSU mouse Megakaryocyte 1 ng RNA-seq   \n",
       "\n",
       "                              Platform Dataset Name  \n",
       "ENCFF547CXK  RNAseq (Illumina NextSeq)       ENCODE  \n",
       "ENCFF774DIF  RNAseq (Illumina NextSeq)       ENCODE  \n",
       "ENCFF155RUP    RNAseq (Illumina HiSeq)       ENCODE  \n",
       "ENCFF702TOW    RNAseq (Illumina HiSeq)       ENCODE  \n",
       "ENCFF963EBQ    RNAseq (Illumina HiSeq)       ENCODE  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_atlas_samples = pd.concat([samples_encode, samples_haemosphere, samples_s4m])\n",
    "print(mouse_atlas_samples.shape)\n",
    "mouse_atlas_samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1074, 5) (14222, 1074)\n"
     ]
    }
   ],
   "source": [
    "# remove duplicated samples that are included in multiple data collection\n",
    "mouse_atlas_samples = mouse_atlas_samples[~mouse_atlas_samples.index.duplicated() & mouse_atlas_samples.index.notna()]\n",
    "mouse_atlas_expression = mouse_atlas_expression.loc[:,~mouse_atlas_expression.columns.duplicated()]\n",
    "print(mouse_atlas_samples.shape, mouse_atlas_expression.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same the integrated expression table and sample table \n",
    "mouse_atlas_expression.to_csv('../data/interim/mouse_atlas/mouse_atlas_expression.tsv', sep='\\t')\n",
    "mouse_atlas_samples.to_csv('../data/interim/mouse_atlas/mouse_atlas_samples.tsv', sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
