{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aim: Collect and combine relevant mouse expression data from the Stemformatics data portal \n",
    "\n",
    "Link: https://www.stemformatics.org/workbench/download_multiple_datasets.\n",
    "\n",
    "Stemformatics is an established gene expression data portal containing over 420 public gene expression datasets derived from microarray, RNA sequencing and single cell profiling technologies. It includes curated ‘collections’ of data relevant to cell reprogramming, as well as hematopoiesis and leukaemia.\n",
    "\n",
    "### Samples\n",
    "\n",
    "Set the serch field to 'species' and use 'Mus musculus' as search key. \n",
    "\n",
    "### Processing steps \n",
    "\n",
    "- Sample selection using samples metadata\n",
    "- Merge RNAseq datasets\n",
    "- Merge microarray (Illumina) datasets\n",
    "- Merge microarray (Affymetrix) datasets \n",
    "- Combine all Stemformatics datasets together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import atlas\n",
    "import plotly.graph_objs as go\n",
    "import handler\n",
    "import requests "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ds_id</th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>replicate_group_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>generic_sample_type</th>\n",
       "      <th>final_cell_type</th>\n",
       "      <th>parental_cell_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>labelling</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chip_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM565763</th>\n",
       "      <td>6146</td>\n",
       "      <td>Polo</td>\n",
       "      <td>Skeletal muscle precursor-derived iPS, replica...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>skeletal muscle precursor-derived iPSC</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>skeletal muscle precursor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM565757</th>\n",
       "      <td>6146</td>\n",
       "      <td>Polo</td>\n",
       "      <td>B cell-derived iPS, replicate 2</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>B cell-derived iPSC</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM565758</th>\n",
       "      <td>6146</td>\n",
       "      <td>Polo</td>\n",
       "      <td>B cell-derived iPS, replicate 3</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>B cell-derived iPSC</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM565756</th>\n",
       "      <td>6146</td>\n",
       "      <td>Polo</td>\n",
       "      <td>B cell-derived iPS, replicate 1</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>B cell-derived iPSC</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM565760</th>\n",
       "      <td>6146</td>\n",
       "      <td>Polo</td>\n",
       "      <td>Tail tip fibroblast-derived iPS, replicate 2</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Tail tip fibroblast-derived iPSC</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ds_id dataset_name  \\\n",
       "chip_id                         \n",
       "GSM565763   6146         Polo   \n",
       "GSM565757   6146         Polo   \n",
       "GSM565758   6146         Polo   \n",
       "GSM565756   6146         Polo   \n",
       "GSM565760   6146         Polo   \n",
       "\n",
       "                                          replicate_group_id      organism  \\\n",
       "chip_id                                                                      \n",
       "GSM565763  Skeletal muscle precursor-derived iPS, replica...  Mus musculus   \n",
       "GSM565757                    B cell-derived iPS, replicate 2  Mus musculus   \n",
       "GSM565758                    B cell-derived iPS, replicate 3  Mus musculus   \n",
       "GSM565756                    B cell-derived iPS, replicate 1  Mus musculus   \n",
       "GSM565760       Tail tip fibroblast-derived iPS, replicate 2  Mus musculus   \n",
       "\n",
       "                                      sample_type  \\\n",
       "chip_id                                             \n",
       "GSM565763  skeletal muscle precursor-derived iPSC   \n",
       "GSM565757                     B cell-derived iPSC   \n",
       "GSM565758                     B cell-derived iPSC   \n",
       "GSM565756                     B cell-derived iPSC   \n",
       "GSM565760        Tail tip fibroblast-derived iPSC   \n",
       "\n",
       "                     generic_sample_type                final_cell_type  \\\n",
       "chip_id                                                                   \n",
       "GSM565763  induced pluripotent stem cell  induced pluripotent stem cell   \n",
       "GSM565757  induced pluripotent stem cell  induced pluripotent stem cell   \n",
       "GSM565758  induced pluripotent stem cell  induced pluripotent stem cell   \n",
       "GSM565756  induced pluripotent stem cell  induced pluripotent stem cell   \n",
       "GSM565760  induced pluripotent stem cell  induced pluripotent stem cell   \n",
       "\n",
       "                  parental_cell_type  sex labelling  \n",
       "chip_id                                              \n",
       "GSM565763  skeletal muscle precursor  NaN       NaN  \n",
       "GSM565757                     B cell  NaN       NaN  \n",
       "GSM565758                     B cell  NaN       NaN  \n",
       "GSM565756                     B cell  NaN       NaN  \n",
       "GSM565760                 fibroblast  NaN       NaN  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# inspect the samples metadata\n",
    "samples = pd.read_csv('/Users/monica/Downloads/export_metadata_samples_v7.2.4.tsv', sep='\\t', index_col=2)\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many of the samples are not healthy blood cells (e.g. iPSCs, AML samples etc.). We will need to select for healthy blood cells from the metadata, and download only the selected samples. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sample selction "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_samples(samples):\n",
    "    '''This function takes the Stemformatics samples metadata and returns samples that are annotated to be blood cells.'''\n",
    "    pos_selected_id = []\n",
    "    neg_selected_id = []\n",
    "    patterns_pos = ['\\WBM', '[Bb]one', 'lymp', '[Hh]aem', '[Hh]em', 'HSC','[Mm]ono', '[Bb]-*\\ *cell', '[Mm]yelo', 'killer',\n",
    "                    'NK', '[Mm]eg', '[Bb]aso', '[Nn]eut', '[Ee]os', '[Pp]las', '[Ee]ryt', '\\W[Tt]-*\\ *cell', 'DC', '[Dd]endri', \n",
    "                    'phage', 'macr']\n",
    "    patterns_neg = ['iPS', 'MSC', 'AML', 'reprogram', 'MAPC', 'KO', 'endothelial', 'LPS', 'mutant', 'Dusp', 'LCMV', 'LSK', 'Chaudhury']\n",
    "    for col in samples.columns:\n",
    "        l = samples[samples[col].notna()]\n",
    "        for p in patterns_pos:\n",
    "            pos_selected_id += l[(l[col].astype(str).str.contains(p) == True)].index.tolist()\n",
    "        for n in patterns_neg: \n",
    "            neg_selected_id += l[(l[col].astype(str).str.contains(n) == True)].index.tolist()\n",
    "    selected = samples.loc[samples.index.isin(set(pos_selected_id))]\n",
    "    return selected.loc[~selected.index.isin(set(neg_selected_id))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(327, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ds_id</th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>replicate_group_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>generic_sample_type</th>\n",
       "      <th>final_cell_type</th>\n",
       "      <th>parental_cell_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>labelling</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chip_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM98876</th>\n",
       "      <td>6658</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM98877</th>\n",
       "      <td>6658</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM98878</th>\n",
       "      <td>6658</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM98882</th>\n",
       "      <td>6658</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC young  animal</td>\n",
       "      <td>LT-HSC young  animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM98879</th>\n",
       "      <td>6658</td>\n",
       "      <td>Rossi</td>\n",
       "      <td>Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>LT-HSC Aged animal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ds_id dataset_name  \\\n",
       "chip_id                        \n",
       "GSM98876   6658        Rossi   \n",
       "GSM98877   6658        Rossi   \n",
       "GSM98878   6658        Rossi   \n",
       "GSM98882   6658        Rossi   \n",
       "GSM98879   6658        Rossi   \n",
       "\n",
       "                                         replicate_group_id      organism  \\\n",
       "chip_id                                                                     \n",
       "GSM98876  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "GSM98877  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "GSM98878  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "GSM98882  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "GSM98879  Lin- c-kit+ Sca1+Flk2-CD34- KLSflk2-CD34-  old...  Mus musculus   \n",
       "\n",
       "                   sample_type   generic_sample_type final_cell_type  \\\n",
       "chip_id                                                                \n",
       "GSM98876    LT-HSC Aged animal    LT-HSC Aged animal             NaN   \n",
       "GSM98877    LT-HSC Aged animal    LT-HSC Aged animal             NaN   \n",
       "GSM98878    LT-HSC Aged animal    LT-HSC Aged animal             NaN   \n",
       "GSM98882  LT-HSC young  animal  LT-HSC young  animal             NaN   \n",
       "GSM98879    LT-HSC Aged animal    LT-HSC Aged animal             NaN   \n",
       "\n",
       "         parental_cell_type  sex labelling  \n",
       "chip_id                                     \n",
       "GSM98876                NaN  NaN       NaN  \n",
       "GSM98877                NaN  NaN       NaN  \n",
       "GSM98878                NaN  NaN       NaN  \n",
       "GSM98882                NaN  NaN       NaN  \n",
       "GSM98879                NaN  NaN       NaN  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_samples = select_samples(samples)\n",
    "print(selected_samples.shape)\n",
    "selected_samples.head() # 327 samples are selected"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine datasets based on platforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ds_id', 'dataset_name', 'replicate_group_id', 'organism',\n",
       "       'sample_type', 'generic_sample_type', 'final_cell_type',\n",
       "       'parental_cell_type', 'sex', 'labelling', 'description', 'platform'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add platform information of to samples metadata\n",
    "datasets = pd.read_csv('/Users/monica/Downloads/export_metadata_datasets_v7.2.4.tsv', sep='\\t', index_col=0)\n",
    "selected_ds = datasets[datasets.index.isin(set(selected_samples.ds_id))]\n",
    "selected_samples = pd.merge(selected_samples, selected_ds[['description', 'platform']], left_on='ds_id', right_index=True)\n",
    "selected_samples.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "platform                                                        ds_id\n",
       "Microarray (Affymetrix MoEx-1_0-ST V1)                          6405      3\n",
       "Microarray (Affymetrix MoGene-1_0-ST V1)                        6264      3\n",
       "                                                                6310     18\n",
       "                                                                6313      4\n",
       "                                                                6455      3\n",
       "                                                                6831     86\n",
       "                                                                7131      8\n",
       "                                                                7269      9\n",
       "Microarray (Affymetrix Mouse430_2 (GPL1261 and A-AFFY-45))      6498      6\n",
       "                                                                6658      8\n",
       "                                                                6659     27\n",
       "                                                                6756     11\n",
       "                                                                6988      2\n",
       "Microarray (Affymetrix Mouse430_2)                              6087      9\n",
       "                                                                6108      4\n",
       "                                                                6300      6\n",
       "Microarray (Agilent 014868 4x44)                                7010      6\n",
       "Microarray (Agilent SurePrint G3 Mouse GE 8x60K 028005)         6290      6\n",
       "                                                                6446      6\n",
       "                                                                6602      3\n",
       "Microarray (GPL81)                                              6109      2\n",
       "Microarray (Illumina Mouse Ref-6 V1)                            6185     18\n",
       "Microarray (Illumina MouseWG-6 v2 (GPL 6887 and A-MEXP-1175))   6637     15\n",
       "                                                                7291     12\n",
       "Microarray (MoGene-2_0-st (GPL16570))                           7211     16\n",
       "RNASeq (Illumina HiSeq 2000 (GPL13112))                         7224      9\n",
       "                                                                7267     12\n",
       "RNASeq (Illumina HiSeq 2000 RNASeq mm9 Gene Counts (GPL13112))  6655     12\n",
       "RNASeq (Illumina MiSeq)                                         6767      3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect the distribution of platforms from which the samples data were generated \n",
    "selected_samples.groupby(['platform', 'ds_id']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group selected ds_ids based on platforms\n",
    "# Note that latforms with not enough dataset representation and small sample size, including Illumina Ref-6, GPL81, MoGene2, \n",
    "# and MoEx1 are excluded. \n",
    "\n",
    "RNAseq_id             = [7224, 7267, 6655, 6767]\n",
    "Agilent_SurePrint_id  = [6290, 6446, 6602]\n",
    "Illu_MouseWG6_id      = [6637, 7291]\n",
    "Affy_Mouse430_id      = [6498, 6658, 6659, 6756, 6988, 6087, 6108, 6300]\n",
    "Affy_MoGene1_id       = [6264, 6310, 6313, 6455, 6831, 7131]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_probes(df, probe_mapping):\n",
    "    '''\n",
    "    Input: A microarray expression dataframe and and a probe mapping table.\n",
    "    Output: The expression dataframe with index changed from probe ids to ensembl gene ids according to the supplied mapping table.\n",
    "    '''\n",
    "    probe_mapping.columns = ['ensembl']\n",
    "    return pd.merge(df, probe_mapping, how='inner', left_index=True, right_index=True).set_index('ensembl')\n",
    "\n",
    "def find_common_genes(dfs):\n",
    "    '''\n",
    "    Input: A list of expression dataframes.\n",
    "    Output: A list of common genes that are appeared in all dataframes.\n",
    "    '''\n",
    "    common_genes = dfs[0].index\n",
    "    for df in dfs:\n",
    "        common_genes = common_genes.intersection(df.index)       \n",
    "    return common_genes\n",
    "\n",
    "def merge_columns(dfs, common_genes):\n",
    "    '''\n",
    "    Input: A list of expression datasts and a list of common genes.\n",
    "    Output: A combined dataframe of all supplied datasets and keep only the common genes. \n",
    "    '''\n",
    "    matrix = []\n",
    "    for df in dfs:\n",
    "        df = df.loc[df.index.isin(common_genes)] # Filter by common genes probed in all datasets \n",
    "        m = df.groupby(df.index).first()          # Resolve rows with duplicate indexes by grouping together with the mean value \n",
    "        matrix.append(m)\n",
    "    return pd.concat(matrix, axis=1)\n",
    "\n",
    "def merge_by_platform(id_list, probe_mapping):\n",
    "    '''\n",
    "    This function takes a list of dataset ids and a probe mapping table as inputs, and will load the corresponding datasets into \n",
    "    variables formatted as dsxxxx with xxxx indicate the ds_id. If the function is supplied with a probe mapping, it will replace \n",
    "    the data index from probe ids to gene ids according to the probe mappintg table.\n",
    "    '''\n",
    "    \n",
    "    serverURL = 'api.stemformatics.org'\n",
    "    headers = {'Content-type': 'application/json'}\n",
    "    load_failed = []\n",
    "    dfs = []\n",
    "    for ds_id in id_list:\n",
    "        try:\n",
    "            result = requests.get('https://%s/expression/%s/raw' % (serverURL, ds_id), headers=headers, verify=False).json()\n",
    "            df = pd.DataFrame(result['data'], index=result['index'], columns=result['columns'])\n",
    "            if probe_mapping is not None:\n",
    "                df = replace_probes(df, probe_mapping) \n",
    "            dfs.append(df)\n",
    "        except ValueError:\n",
    "            load_failed.append(ds_id)\n",
    "            continue \n",
    "    print('Successfully load datasets %s' % [i for i in id_list if i not in load_failed])\n",
    "    return pd.concat(dfs, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully load datasets [7224, 7267, 6767]\n",
      "Successfully load datasets [6290, 6446, 6602]\n",
      "Successfully load datasets [6637]\n",
      "Successfully load datasets [6498, 6658, 6659, 6756, 6988, 6108, 6300]\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "def combine_s4m_ds():   \n",
    "    '''\n",
    "    This function will combine all the selected datasets form the stemformatics portal. \n",
    "    '''\n",
    "    # Load the selected s4m datasets from the data portal api and replace the probe ID with gene ID\n",
    "    genes_WG6        = pd.read_csv('../data/raw/Stemformatics/probes_MouseWG6_v2.tsv', sep='\\t', index_col = 0, header = None)\n",
    "    genes_Mouse430_2 = pd.read_csv('../data/raw/Stemformatics/probes_Mouse430_2.tsv', sep='\\t', index_col = 0, header = None)\n",
    "    genes_MoGene1    = pd.read_csv('../data/raw/Stemformatics/probes_MoGene1.tsv', sep='\\t', index_col = 0, header = None)\n",
    "    genes_SurePrint  = pd.read_csv('../data/raw/Stemformatics/probes_SurePrintG3.tsv', sep='\\t', index_col = 0, header = None)\n",
    "    \n",
    "    \n",
    "    df_RNAseq    = merge_by_platform(RNAseq_id, None)\n",
    "    df_SurePrint = merge_by_platform(Agilent_SurePrint_id, genes_SurePrint)\n",
    "    df_WG6       = merge_by_platform(Illu_MouseWG6_id, genes_WG6)\n",
    "    df_Mouse430  = merge_by_platform(Affy_Mouse430_id, genes_Mouse430_2)\n",
    "    df_MoGene1   = merge_by_platform(Affy_MoGene1_id, genes_MoGene1)\n",
    "    \n",
    "    df_Mouse430.columns = df_Mouse430.columns.str.strip('.CEL')\n",
    "    df_MoGene1.columns = df_MoGene1.columns.str.strip('.CEL')\n",
    "    \n",
    "    dfs = [df_RNAseq, df_SurePrint, df_WG6, df_Mouse430, df_MoGene1]\n",
    "    \n",
    "    common_genes = find_common_genes(dfs)\n",
    "    combined_df = merge_columns(dfs, common_genes)\n",
    "    output = combined_df.loc[:,combined_df.columns.isin(selected_samples.index)]\n",
    "    return atlas.rankTransform(output)\n",
    "\n",
    "expression_s4m = combine_s4m_ds()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17947, 240)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ERR1308082</th>\n",
       "      <th>ERR1308083</th>\n",
       "      <th>ERR1308084</th>\n",
       "      <th>ERR1308085</th>\n",
       "      <th>ERR1308086</th>\n",
       "      <th>ERR1308087</th>\n",
       "      <th>ERR1308088</th>\n",
       "      <th>ERR1308089</th>\n",
       "      <th>ERR1308090</th>\n",
       "      <th>GSM2631976</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM854333</th>\n",
       "      <th>GSM854334</th>\n",
       "      <th>GSM1023629</th>\n",
       "      <th>GSM1023630</th>\n",
       "      <th>GSM1023631</th>\n",
       "      <th>GSM1023632</th>\n",
       "      <th>GSM1023633</th>\n",
       "      <th>GSM1023634</th>\n",
       "      <th>GSM1023635</th>\n",
       "      <th>GSM1023636</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.945896</td>\n",
       "      <td>0.906920</td>\n",
       "      <td>0.882738</td>\n",
       "      <td>0.952193</td>\n",
       "      <td>0.781551</td>\n",
       "      <td>0.905834</td>\n",
       "      <td>0.940380</td>\n",
       "      <td>0.909790</td>\n",
       "      <td>0.924946</td>\n",
       "      <td>0.944392</td>\n",
       "      <td>...</td>\n",
       "      <td>0.863153</td>\n",
       "      <td>0.854237</td>\n",
       "      <td>0.931186</td>\n",
       "      <td>0.934195</td>\n",
       "      <td>0.925169</td>\n",
       "      <td>0.926060</td>\n",
       "      <td>0.922104</td>\n",
       "      <td>0.925781</td>\n",
       "      <td>0.939544</td>\n",
       "      <td>0.938653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.137850</td>\n",
       "      <td>0.160305</td>\n",
       "      <td>0.105644</td>\n",
       "      <td>0.136346</td>\n",
       "      <td>0.155402</td>\n",
       "      <td>0.151251</td>\n",
       "      <td>0.269181</td>\n",
       "      <td>0.076392</td>\n",
       "      <td>0.091074</td>\n",
       "      <td>0.175350</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001839</td>\n",
       "      <td>0.002563</td>\n",
       "      <td>0.008692</td>\n",
       "      <td>0.004736</td>\n",
       "      <td>0.003287</td>\n",
       "      <td>0.010030</td>\n",
       "      <td>0.008692</td>\n",
       "      <td>0.008748</td>\n",
       "      <td>0.003677</td>\n",
       "      <td>0.005126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.557558</td>\n",
       "      <td>0.592690</td>\n",
       "      <td>0.604196</td>\n",
       "      <td>0.490444</td>\n",
       "      <td>0.652950</td>\n",
       "      <td>0.533320</td>\n",
       "      <td>0.711428</td>\n",
       "      <td>0.645623</td>\n",
       "      <td>0.752103</td>\n",
       "      <td>0.539449</td>\n",
       "      <td>...</td>\n",
       "      <td>0.761854</td>\n",
       "      <td>0.770714</td>\n",
       "      <td>0.776174</td>\n",
       "      <td>0.842313</td>\n",
       "      <td>0.803923</td>\n",
       "      <td>0.886945</td>\n",
       "      <td>0.878476</td>\n",
       "      <td>0.869449</td>\n",
       "      <td>0.852844</td>\n",
       "      <td>0.702959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>0.361370</td>\n",
       "      <td>0.297431</td>\n",
       "      <td>0.105644</td>\n",
       "      <td>0.136346</td>\n",
       "      <td>0.115758</td>\n",
       "      <td>0.151251</td>\n",
       "      <td>0.250571</td>\n",
       "      <td>0.076392</td>\n",
       "      <td>0.091074</td>\n",
       "      <td>0.361843</td>\n",
       "      <td>...</td>\n",
       "      <td>0.263387</td>\n",
       "      <td>0.177411</td>\n",
       "      <td>0.942720</td>\n",
       "      <td>0.914415</td>\n",
       "      <td>0.866329</td>\n",
       "      <td>0.780910</td>\n",
       "      <td>0.825988</td>\n",
       "      <td>0.857135</td>\n",
       "      <td>0.079122</td>\n",
       "      <td>0.115005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.484287</td>\n",
       "      <td>0.469159</td>\n",
       "      <td>0.387892</td>\n",
       "      <td>0.491196</td>\n",
       "      <td>0.416588</td>\n",
       "      <td>0.450326</td>\n",
       "      <td>0.450047</td>\n",
       "      <td>0.403382</td>\n",
       "      <td>0.511673</td>\n",
       "      <td>0.175350</td>\n",
       "      <td>...</td>\n",
       "      <td>0.047139</td>\n",
       "      <td>0.079289</td>\n",
       "      <td>0.334513</td>\n",
       "      <td>0.484594</td>\n",
       "      <td>0.398702</td>\n",
       "      <td>0.331922</td>\n",
       "      <td>0.415390</td>\n",
       "      <td>0.358500</td>\n",
       "      <td>0.226890</td>\n",
       "      <td>0.373656</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 240 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ERR1308082  ERR1308083  ERR1308084  ERR1308085  \\\n",
       "ensembl                                                              \n",
       "ENSMUSG00000000001    0.945896    0.906920    0.882738    0.952193   \n",
       "ENSMUSG00000000003    0.137850    0.160305    0.105644    0.136346   \n",
       "ENSMUSG00000000028    0.557558    0.592690    0.604196    0.490444   \n",
       "ENSMUSG00000000031    0.361370    0.297431    0.105644    0.136346   \n",
       "ENSMUSG00000000037    0.484287    0.469159    0.387892    0.491196   \n",
       "\n",
       "                    ERR1308086  ERR1308087  ERR1308088  ERR1308089  \\\n",
       "ensembl                                                              \n",
       "ENSMUSG00000000001    0.781551    0.905834    0.940380    0.909790   \n",
       "ENSMUSG00000000003    0.155402    0.151251    0.269181    0.076392   \n",
       "ENSMUSG00000000028    0.652950    0.533320    0.711428    0.645623   \n",
       "ENSMUSG00000000031    0.115758    0.151251    0.250571    0.076392   \n",
       "ENSMUSG00000000037    0.416588    0.450326    0.450047    0.403382   \n",
       "\n",
       "                    ERR1308090  GSM2631976  ...  GSM854333  GSM854334  \\\n",
       "ensembl                                     ...                         \n",
       "ENSMUSG00000000001    0.924946    0.944392  ...   0.863153   0.854237   \n",
       "ENSMUSG00000000003    0.091074    0.175350  ...   0.001839   0.002563   \n",
       "ENSMUSG00000000028    0.752103    0.539449  ...   0.761854   0.770714   \n",
       "ENSMUSG00000000031    0.091074    0.361843  ...   0.263387   0.177411   \n",
       "ENSMUSG00000000037    0.511673    0.175350  ...   0.047139   0.079289   \n",
       "\n",
       "                    GSM1023629  GSM1023630  GSM1023631  GSM1023632  \\\n",
       "ensembl                                                              \n",
       "ENSMUSG00000000001    0.931186    0.934195    0.925169    0.926060   \n",
       "ENSMUSG00000000003    0.008692    0.004736    0.003287    0.010030   \n",
       "ENSMUSG00000000028    0.776174    0.842313    0.803923    0.886945   \n",
       "ENSMUSG00000000031    0.942720    0.914415    0.866329    0.780910   \n",
       "ENSMUSG00000000037    0.334513    0.484594    0.398702    0.331922   \n",
       "\n",
       "                    GSM1023633  GSM1023634  GSM1023635  GSM1023636  \n",
       "ensembl                                                             \n",
       "ENSMUSG00000000001    0.922104    0.925781    0.939544    0.938653  \n",
       "ENSMUSG00000000003    0.008692    0.008748    0.003677    0.005126  \n",
       "ENSMUSG00000000028    0.878476    0.869449    0.852844    0.702959  \n",
       "ENSMUSG00000000031    0.825988    0.857135    0.079122    0.115005  \n",
       "ENSMUSG00000000037    0.415390    0.358500    0.226890    0.373656  \n",
       "\n",
       "[5 rows x 240 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(expression_s4m.shape)\n",
    "expression_s4m.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Since some of the selected datasets failed to be downloaded from the stemformatics api, we will need to select the samples metadata \n",
    "# by the samples that are actually downloaded. \n",
    "samples_s4m = selected_samples[selected_samples.index.isin(expression_s4m.columns)]\n",
    "samples_s4m.shape\n",
    "samples_s4m.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the combined expression data and samples metadata\n",
    "expression_s4m.to_csv('../data/interim/mouse_integrate/expression_s4m.tsv', sep='\\t')\n",
    "samples_s4m.to_csv('../data/interim/mouse_integrate/samples_s4m.tsv', sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
