{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aim: Collect and combine relevant mouse expression data from the Haemosphere data portal \n",
    "\n",
    "Link: https://haemosphere.org/datasets/show\n",
    "\n",
    "Haemosphere is an online data portal that contains many gene expression datasets that focused on haematopoeisis in mice and humans. \n",
    "\n",
    "### Datasets \n",
    "\n",
    "We selected 6 datasets:\n",
    "\n",
    "|__No.__|__dataset name__|__platform__|__platform details__|\n",
    "|-------|----------------|------------|--------------------|\n",
    "|1|Haemopedia-Mouse-RNAseq|RNAseq|HiSeq|\n",
    "|2|Immgen-ULI-RNAseq|RNAseq|Illumina NextSeq 500|\n",
    "|3|Haemopedia|microarray|Illumina WG6|\n",
    "|4|Gene-Expression-Commons|microarray|Affymetrix Mouse430_2|\n",
    "|5|Immgen|microarray|Affymetrix Mouse Gene 1.0 ST Array|\n",
    "|6|Goodell|microarray|Affymetrix Mouse430_2|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import atlas\n",
    "import handler\n",
    "import GEOparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Load data\n",
    "\n",
    "# Haemopedia RNAseq dataset \n",
    "expression1 = pd.read_csv('../data/raw/Haemosphere/Haemopedia-Mouse-RNASeq/Haemopedia-Mouse-RNASeq_tpm.txt', sep='\\t', index_col=0)\n",
    "samples1    = pd.read_csv('../data/raw/Haemosphere/Haemopedia-Mouse-RNASeq/Haemopedia-Mouse-RNASeq_samples.txt', sep='\\t', index_col=0)\n",
    "\n",
    "# Immgen ULI dataset \n",
    "expression2 = pd.read_csv('../data/raw/Haemosphere/Immgen-RNAseq/Immgen-ULI-RNASeq_tpm.txt', sep='\\t', index_col=0)\n",
    "samples2    = pd.read_csv('../data/raw/Haemosphere/Immgen-RNAseq/Immgen-ULI-RNASeq_samples.txt', sep='\\t', index_col=0)\n",
    "\n",
    "# Haemopedia microarray dataset \n",
    "expression3 = pd.read_csv('../data/raw/Haemosphere/Haemopedia-microarray/Haemopedia_expression.txt', sep='\\t', index_col=0)\n",
    "samples3    = pd.read_csv('../data/raw/Haemosphere/Haemopedia-microarray/Haemopedia_samples.txt', sep='\\t', index_col=0)\n",
    "probes3     = pd.read_csv('../data/raw/Haemosphere/Haemopedia-microarray/Haemopedia_probes.txt', sep='\\t', index_col=0)\n",
    "\n",
    "# Gene expression commons dataset\n",
    "expression4 = pd.read_csv('../data/raw/Haemosphere/Gene-Expression-Commons/Gene-Expression-Commons_expression.txt', sep='\\t', index_col=0)\n",
    "samples4    = pd.read_csv('../data/raw/Haemosphere/Gene-Expression-Commons/Gene-Expression-Commons_samples.txt', sep='\\t', index_col=0)\n",
    "probes4     = pd.read_csv('../data/raw/Haemosphere/Gene-Expression-Commons/Gene-Expression-Commons_probes.txt', sep='\\t', index_col=0)\n",
    "\n",
    "# Immgen microarray dataset\n",
    "expression5 = pd.read_csv('../data/raw/Haemosphere/Immgen-microarray/Immgen_expression.txt', sep='\\t', index_col=0)\n",
    "samples5    = pd.read_csv('../data/raw/Haemosphere/Immgen-microarray/Immgen_samples.txt', sep='\\t', index_col=0)\n",
    "probes5     = pd.read_csv('../data/raw/Haemosphere/Immgen-microarray/Immgen_probes.txt', sep='\\t', index_col=0)\n",
    "\n",
    "# Goodell dataset\n",
    "expression6 = pd.read_csv('../data/raw/Haemosphere/Goodell-microarray/Goodell_expression.txt', sep='\\t', index_col=0)\n",
    "samples6    = pd.read_csv('../data/raw/Haemosphere/Goodell-microarray/Goodell_samples.txt', sep='\\t', index_col=0)\n",
    "probes6     = pd.read_csv('../data/raw/Haemosphere/Goodell-microarray/Goodell_probes.txt', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For sample metadata contains standardised sample ids such as GEO_accessions and \n",
    "# SRA_accessions, use them as sampleId.\n",
    "expression1 = pd.merge(expression1.T, samples1['sampleCode'], left_index=True, right_index=True).set_index('sampleCode').T\n",
    "samples1 = samples1.set_index('sampleCode')\n",
    "\n",
    "expression2 = pd.merge(expression2.T, samples2['sample_geo_accession'], left_index=True, right_index=True).set_index('sample_geo_accession').T\n",
    "samples2 = samples2.set_index('sample_geo_accession')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge selected expression datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_probes(df, probe_mapping):\n",
    "    '''\n",
    "    Input: A microarray expression dataframe and and a probe mapping table.\n",
    "    Output: The expression dataframe with index changed from probe ids to ensembl gene ids according to the supplied mapping table.\n",
    "    '''\n",
    "    probe_mapping.columns = ['ensembl']\n",
    "    return pd.merge(df, probe_mapping, how='inner', left_index=True, right_index=True).set_index('ensembl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For microarray datasets, replace index of expression table from probe id to gene id\n",
    "expression3 = replace_probes(expression3, probes3)\n",
    "expression4 = replace_probes(expression4, probes4)\n",
    "expression5 = replace_probes(expression5, probes5)\n",
    "expression6 = replace_probes(expression6, probes6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14462, 806)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAGRF7658.1338</th>\n",
       "      <th>CAGRF7658.1339</th>\n",
       "      <th>CAGRF7126.1213</th>\n",
       "      <th>CAGRF7126.1217</th>\n",
       "      <th>CAGRF7126.1222</th>\n",
       "      <th>CAGRF7126.1223</th>\n",
       "      <th>CAGRF7126.1229</th>\n",
       "      <th>CAGRF7126.1232</th>\n",
       "      <th>CAGRF7126.1249</th>\n",
       "      <th>CAGRF7126.1250</th>\n",
       "      <th>...</th>\n",
       "      <th>T-CD8-N(1)</th>\n",
       "      <th>T-CD8-N(2)</th>\n",
       "      <th>B-Cell(1)</th>\n",
       "      <th>B-Cell(2)</th>\n",
       "      <th>Monocyte(1)</th>\n",
       "      <th>Monocyte(2)</th>\n",
       "      <th>Granulocyte(1)</th>\n",
       "      <th>Granulocyte(2)</th>\n",
       "      <th>N-Er(1)</th>\n",
       "      <th>N-Er(2)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.954778</td>\n",
       "      <td>0.957129</td>\n",
       "      <td>0.971719</td>\n",
       "      <td>0.955193</td>\n",
       "      <td>0.939566</td>\n",
       "      <td>0.943300</td>\n",
       "      <td>0.950975</td>\n",
       "      <td>0.968815</td>\n",
       "      <td>0.983128</td>\n",
       "      <td>0.984857</td>\n",
       "      <td>...</td>\n",
       "      <td>0.930300</td>\n",
       "      <td>0.921864</td>\n",
       "      <td>0.897490</td>\n",
       "      <td>0.921726</td>\n",
       "      <td>0.944199</td>\n",
       "      <td>0.973724</td>\n",
       "      <td>0.986032</td>\n",
       "      <td>0.971270</td>\n",
       "      <td>0.858699</td>\n",
       "      <td>0.891440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.093521</td>\n",
       "      <td>0.098188</td>\n",
       "      <td>0.123773</td>\n",
       "      <td>0.124948</td>\n",
       "      <td>0.134387</td>\n",
       "      <td>0.129581</td>\n",
       "      <td>0.126089</td>\n",
       "      <td>0.104135</td>\n",
       "      <td>0.145346</td>\n",
       "      <td>0.143341</td>\n",
       "      <td>...</td>\n",
       "      <td>0.267356</td>\n",
       "      <td>0.267218</td>\n",
       "      <td>0.264383</td>\n",
       "      <td>0.269707</td>\n",
       "      <td>0.262308</td>\n",
       "      <td>0.273579</td>\n",
       "      <td>0.257744</td>\n",
       "      <td>0.262965</td>\n",
       "      <td>0.270882</td>\n",
       "      <td>0.269292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.708823</td>\n",
       "      <td>0.723621</td>\n",
       "      <td>0.511686</td>\n",
       "      <td>0.907343</td>\n",
       "      <td>0.934656</td>\n",
       "      <td>0.932859</td>\n",
       "      <td>0.899253</td>\n",
       "      <td>0.854930</td>\n",
       "      <td>0.763726</td>\n",
       "      <td>0.768842</td>\n",
       "      <td>...</td>\n",
       "      <td>0.536475</td>\n",
       "      <td>0.535299</td>\n",
       "      <td>0.540140</td>\n",
       "      <td>0.451148</td>\n",
       "      <td>0.572293</td>\n",
       "      <td>0.564963</td>\n",
       "      <td>0.669686</td>\n",
       "      <td>0.626469</td>\n",
       "      <td>0.871802</td>\n",
       "      <td>0.867584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.359217</td>\n",
       "      <td>0.366616</td>\n",
       "      <td>0.123773</td>\n",
       "      <td>0.562370</td>\n",
       "      <td>0.581939</td>\n",
       "      <td>0.578689</td>\n",
       "      <td>0.541419</td>\n",
       "      <td>0.463698</td>\n",
       "      <td>0.323192</td>\n",
       "      <td>0.143341</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121387</td>\n",
       "      <td>0.121871</td>\n",
       "      <td>0.122493</td>\n",
       "      <td>0.122943</td>\n",
       "      <td>0.125640</td>\n",
       "      <td>0.121007</td>\n",
       "      <td>0.124188</td>\n",
       "      <td>0.121387</td>\n",
       "      <td>0.501175</td>\n",
       "      <td>0.402088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>0.358249</td>\n",
       "      <td>0.351680</td>\n",
       "      <td>0.123773</td>\n",
       "      <td>0.124948</td>\n",
       "      <td>0.134387</td>\n",
       "      <td>0.129581</td>\n",
       "      <td>0.126089</td>\n",
       "      <td>0.104135</td>\n",
       "      <td>0.145346</td>\n",
       "      <td>0.143341</td>\n",
       "      <td>...</td>\n",
       "      <td>0.436904</td>\n",
       "      <td>0.370868</td>\n",
       "      <td>0.377092</td>\n",
       "      <td>0.383695</td>\n",
       "      <td>0.378959</td>\n",
       "      <td>0.374015</td>\n",
       "      <td>0.358664</td>\n",
       "      <td>0.367065</td>\n",
       "      <td>0.390022</td>\n",
       "      <td>0.381275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 806 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CAGRF7658.1338  CAGRF7658.1339  CAGRF7126.1213  \\\n",
       "ensembl                                                              \n",
       "ENSMUSG00000000001        0.954778        0.957129        0.971719   \n",
       "ENSMUSG00000000003        0.093521        0.098188        0.123773   \n",
       "ENSMUSG00000000028        0.708823        0.723621        0.511686   \n",
       "ENSMUSG00000000037        0.359217        0.366616        0.123773   \n",
       "ENSMUSG00000000049        0.358249        0.351680        0.123773   \n",
       "\n",
       "                    CAGRF7126.1217  CAGRF7126.1222  CAGRF7126.1223  \\\n",
       "ensembl                                                              \n",
       "ENSMUSG00000000001        0.955193        0.939566        0.943300   \n",
       "ENSMUSG00000000003        0.124948        0.134387        0.129581   \n",
       "ENSMUSG00000000028        0.907343        0.934656        0.932859   \n",
       "ENSMUSG00000000037        0.562370        0.581939        0.578689   \n",
       "ENSMUSG00000000049        0.124948        0.134387        0.129581   \n",
       "\n",
       "                    CAGRF7126.1229  CAGRF7126.1232  CAGRF7126.1249  \\\n",
       "ensembl                                                              \n",
       "ENSMUSG00000000001        0.950975        0.968815        0.983128   \n",
       "ENSMUSG00000000003        0.126089        0.104135        0.145346   \n",
       "ENSMUSG00000000028        0.899253        0.854930        0.763726   \n",
       "ENSMUSG00000000037        0.541419        0.463698        0.323192   \n",
       "ENSMUSG00000000049        0.126089        0.104135        0.145346   \n",
       "\n",
       "                    CAGRF7126.1250  ...  T-CD8-N(1)  T-CD8-N(2)  B-Cell(1)  \\\n",
       "ensembl                             ...                                      \n",
       "ENSMUSG00000000001        0.984857  ...    0.930300    0.921864   0.897490   \n",
       "ENSMUSG00000000003        0.143341  ...    0.267356    0.267218   0.264383   \n",
       "ENSMUSG00000000028        0.768842  ...    0.536475    0.535299   0.540140   \n",
       "ENSMUSG00000000037        0.143341  ...    0.121387    0.121871   0.122493   \n",
       "ENSMUSG00000000049        0.143341  ...    0.436904    0.370868   0.377092   \n",
       "\n",
       "                    B-Cell(2)  Monocyte(1)  Monocyte(2)  Granulocyte(1)  \\\n",
       "ensembl                                                                   \n",
       "ENSMUSG00000000001   0.921726     0.944199     0.973724        0.986032   \n",
       "ENSMUSG00000000003   0.269707     0.262308     0.273579        0.257744   \n",
       "ENSMUSG00000000028   0.451148     0.572293     0.564963        0.669686   \n",
       "ENSMUSG00000000037   0.122943     0.125640     0.121007        0.124188   \n",
       "ENSMUSG00000000049   0.383695     0.378959     0.374015        0.358664   \n",
       "\n",
       "                    Granulocyte(2)   N-Er(1)   N-Er(2)  \n",
       "ensembl                                                 \n",
       "ENSMUSG00000000001        0.971270  0.858699  0.891440  \n",
       "ENSMUSG00000000003        0.262965  0.270882  0.269292  \n",
       "ENSMUSG00000000028        0.626469  0.871802  0.867584  \n",
       "ENSMUSG00000000037        0.121387  0.501175  0.402088  \n",
       "ENSMUSG00000000049        0.367065  0.390022  0.381275  \n",
       "\n",
       "[5 rows x 806 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge samples of the 6 selected datasets\n",
    "dfs = [expression1, expression2, expression3, expression4, expression5, expression6]\n",
    "common_genes = handler.find_common_genes(dfs)\n",
    "merged_ds = handler.merge_columns(dfs, common_genes)\n",
    "expression_Haemosphere = atlas.rankTransform(merged_ds)\n",
    "\n",
    "print(expression_Haemosphere.shape)\n",
    "expression_Haemosphere.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Standardise the metadata of each dataset\n",
    "\n",
    "Since the sample metadata of different datasets may have recorded differnet variables, or using different names for similar attributes, to merge them, we will need to identify a minimal list of necessary attributes to describe samples, and unify the naming of these attributes. The essential list of attributes includes:\n",
    "\n",
    "\n",
    "- cell type\n",
    "- cell lineage \n",
    "- description \n",
    "- markers \n",
    "- dataset_name  \n",
    "- platform  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell Type Description</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>tissue</th>\n",
       "      <th>treatment</th>\n",
       "      <th>immunophenotype</th>\n",
       "      <th>species</th>\n",
       "      <th>strain</th>\n",
       "      <th>RNA.Id</th>\n",
       "      <th>batch</th>\n",
       "      <th>firstPublished</th>\n",
       "      <th>cellOrder</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleCode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561641</th>\n",
       "      <td>Fob</td>\n",
       "      <td>Follicular B cells</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>none</td>\n",
       "      <td>FSC-lo B220:+ CD21:+ CD23:+</td>\n",
       "      <td>Murine</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>SRR1561641</td>\n",
       "      <td>GSE60927</td>\n",
       "      <td>Shi et al, 2015; 25894659</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561642</th>\n",
       "      <td>Fob</td>\n",
       "      <td>Follicular B cells</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>none</td>\n",
       "      <td>FSC-lo B220:+ CD21:+ CD23:+</td>\n",
       "      <td>Murine</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>SRR1561642</td>\n",
       "      <td>GSE60927</td>\n",
       "      <td>Shi et al, 2015; 25894659</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561645</th>\n",
       "      <td>MZB</td>\n",
       "      <td>Marginal zone B cells</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>none</td>\n",
       "      <td>FSC-lo B220:+ CD21:-hi CD23:-</td>\n",
       "      <td>Murine</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>SRR1561645</td>\n",
       "      <td>GSE60927</td>\n",
       "      <td>Shi et al, 2015; 25894659</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561646</th>\n",
       "      <td>MZB</td>\n",
       "      <td>Marginal zone B cells</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>none</td>\n",
       "      <td>FSC-lo B220:+ CD21:-hi CD23:-</td>\n",
       "      <td>Murine</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>SRR1561646</td>\n",
       "      <td>GSE60927</td>\n",
       "      <td>Shi et al, 2015; 25894659</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561643</th>\n",
       "      <td>B1Pt</td>\n",
       "      <td>Peritoneal B1 cells</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Peritoneal lavage</td>\n",
       "      <td>none</td>\n",
       "      <td>B220:+ Mac1:-lo CD23:-</td>\n",
       "      <td>Murine</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>SRR1561643</td>\n",
       "      <td>GSE60927</td>\n",
       "      <td>Shi et al, 2015; 25894659</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    celltype  cell Type Description    cell_lineage  \\\n",
       "sampleCode                                                            \n",
       "GSE60927.SRR1561641      Fob     Follicular B cells  B Cell Lineage   \n",
       "GSE60927.SRR1561642      Fob     Follicular B cells  B Cell Lineage   \n",
       "GSE60927.SRR1561645      MZB  Marginal zone B cells  B Cell Lineage   \n",
       "GSE60927.SRR1561646      MZB  Marginal zone B cells  B Cell Lineage   \n",
       "GSE60927.SRR1561643     B1Pt    Peritoneal B1 cells  B Cell Lineage   \n",
       "\n",
       "                                tissue treatment  \\\n",
       "sampleCode                                         \n",
       "GSE60927.SRR1561641             Spleen      none   \n",
       "GSE60927.SRR1561642             Spleen      none   \n",
       "GSE60927.SRR1561645             Spleen      none   \n",
       "GSE60927.SRR1561646             Spleen      none   \n",
       "GSE60927.SRR1561643  Peritoneal lavage      none   \n",
       "\n",
       "                                   immunophenotype species   strain  \\\n",
       "sampleCode                                                            \n",
       "GSE60927.SRR1561641    FSC-lo B220:+ CD21:+ CD23:+  Murine  C57BL/6   \n",
       "GSE60927.SRR1561642    FSC-lo B220:+ CD21:+ CD23:+  Murine  C57BL/6   \n",
       "GSE60927.SRR1561645  FSC-lo B220:+ CD21:-hi CD23:-  Murine  C57BL/6   \n",
       "GSE60927.SRR1561646  FSC-lo B220:+ CD21:-hi CD23:-  Murine  C57BL/6   \n",
       "GSE60927.SRR1561643         B220:+ Mac1:-lo CD23:-  Murine  C57BL/6   \n",
       "\n",
       "                         RNA.Id     batch             firstPublished  \\\n",
       "sampleCode                                                             \n",
       "GSE60927.SRR1561641  SRR1561641  GSE60927  Shi et al, 2015; 25894659   \n",
       "GSE60927.SRR1561642  SRR1561642  GSE60927  Shi et al, 2015; 25894659   \n",
       "GSE60927.SRR1561645  SRR1561645  GSE60927  Shi et al, 2015; 25894659   \n",
       "GSE60927.SRR1561646  SRR1561646  GSE60927  Shi et al, 2015; 25894659   \n",
       "GSE60927.SRR1561643  SRR1561643  GSE60927  Shi et al, 2015; 25894659   \n",
       "\n",
       "                     cellOrder  \n",
       "sampleCode                      \n",
       "GSE60927.SRR1561641          1  \n",
       "GSE60927.SRR1561642          1  \n",
       "GSE60927.SRR1561645          2  \n",
       "GSE60927.SRR1561646          2  \n",
       "GSE60927.SRR1561643          3  "
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Haemopedia RNAseq dataset\n",
    "samples1.head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>description</th>\n",
       "      <th>markers</th>\n",
       "      <th>platform</th>\n",
       "      <th>dataset_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleCode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561641</th>\n",
       "      <td>Fob</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Follicular B cells</td>\n",
       "      <td>FSC-lo B220:+ CD21:+ CD23:+</td>\n",
       "      <td>Illumina HiSeq</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561642</th>\n",
       "      <td>Fob</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Follicular B cells</td>\n",
       "      <td>FSC-lo B220:+ CD21:+ CD23:+</td>\n",
       "      <td>Illumina HiSeq</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561645</th>\n",
       "      <td>MZB</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Marginal zone B cells</td>\n",
       "      <td>FSC-lo B220:+ CD21:-hi CD23:-</td>\n",
       "      <td>Illumina HiSeq</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561646</th>\n",
       "      <td>MZB</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Marginal zone B cells</td>\n",
       "      <td>FSC-lo B220:+ CD21:-hi CD23:-</td>\n",
       "      <td>Illumina HiSeq</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE60927.SRR1561643</th>\n",
       "      <td>B1Pt</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Peritoneal B1 cells</td>\n",
       "      <td>B220:+ Mac1:-lo CD23:-</td>\n",
       "      <td>Illumina HiSeq</td>\n",
       "      <td>Haemopedia-Mouse-RNAseq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    celltype    cell_lineage            description  \\\n",
       "sampleCode                                                            \n",
       "GSE60927.SRR1561641      Fob  B Cell Lineage     Follicular B cells   \n",
       "GSE60927.SRR1561642      Fob  B Cell Lineage     Follicular B cells   \n",
       "GSE60927.SRR1561645      MZB  B Cell Lineage  Marginal zone B cells   \n",
       "GSE60927.SRR1561646      MZB  B Cell Lineage  Marginal zone B cells   \n",
       "GSE60927.SRR1561643     B1Pt  B Cell Lineage    Peritoneal B1 cells   \n",
       "\n",
       "                                           markers        platform  \\\n",
       "sampleCode                                                           \n",
       "GSE60927.SRR1561641    FSC-lo B220:+ CD21:+ CD23:+  Illumina HiSeq   \n",
       "GSE60927.SRR1561642    FSC-lo B220:+ CD21:+ CD23:+  Illumina HiSeq   \n",
       "GSE60927.SRR1561645  FSC-lo B220:+ CD21:-hi CD23:-  Illumina HiSeq   \n",
       "GSE60927.SRR1561646  FSC-lo B220:+ CD21:-hi CD23:-  Illumina HiSeq   \n",
       "GSE60927.SRR1561643         B220:+ Mac1:-lo CD23:-  Illumina HiSeq   \n",
       "\n",
       "                                dataset_name  \n",
       "sampleCode                                    \n",
       "GSE60927.SRR1561641  Haemopedia-Mouse-RNAseq  \n",
       "GSE60927.SRR1561642  Haemopedia-Mouse-RNAseq  \n",
       "GSE60927.SRR1561645  Haemopedia-Mouse-RNAseq  \n",
       "GSE60927.SRR1561646  Haemopedia-Mouse-RNAseq  \n",
       "GSE60927.SRR1561643  Haemopedia-Mouse-RNAseq  "
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Haemopedia RNAseq dataset - after standardisation\n",
    "samples1 = samples1[['celltype', 'cell_lineage', 'cell Type Description', 'immunophenotype']]\n",
    "samples1.columns = ['celltype', 'cell_lineage', 'description', 'markers']\n",
    "samples1['platform'] = 'RNAdeq (Illumina HiSeq)'\n",
    "samples1['dataset_name'] = 'Haemopedia-Mouse-RNAseq'\n",
    "samples1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>organism</th>\n",
       "      <th>strain</th>\n",
       "      <th>age</th>\n",
       "      <th>sort markers</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2932488</th>\n",
       "      <td>B.Fem.Sp</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>6 weeks old</td>\n",
       "      <td>CD19+Igm+TCRb-</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Female Splenic  B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932489</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>6 weeks old</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932490</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>6 weeks old</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932491</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>6 weeks old</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932492</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>C57BL/6J</td>\n",
       "      <td>6 weeks old</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      celltype      organism    strain          age  \\\n",
       "sample_geo_accession                                                  \n",
       "GSM2932488            B.Fem.Sp  Mus musculus  C57BL/6J  6 weeks old   \n",
       "GSM2932489             B.Fo.Sp  Mus musculus  C57BL/6J  6 weeks old   \n",
       "GSM2932490             B.Fo.Sp  Mus musculus  C57BL/6J  6 weeks old   \n",
       "GSM2932491             B.Fo.Sp  Mus musculus  C57BL/6J  6 weeks old   \n",
       "GSM2932492             B.Fo.Sp  Mus musculus  C57BL/6J  6 weeks old   \n",
       "\n",
       "                                            sort markers  tissue  \\\n",
       "sample_geo_accession                                               \n",
       "GSM2932488                                CD19+Igm+TCRb-  Spleen   \n",
       "GSM2932489            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Spleen   \n",
       "GSM2932490            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Spleen   \n",
       "GSM2932491            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Spleen   \n",
       "GSM2932492            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Spleen   \n",
       "\n",
       "                        cell_lineage              description  \n",
       "sample_geo_accession                                           \n",
       "GSM2932488            B Cell Lineage  Female Splenic  B cells  \n",
       "GSM2932489            B Cell Lineage       Splenic Follicular  \n",
       "GSM2932490            B Cell Lineage       Splenic Follicular  \n",
       "GSM2932491            B Cell Lineage       Splenic Follicular  \n",
       "GSM2932492            B Cell Lineage       Splenic Follicular  "
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Immgen ULI RNAseq dataset\n",
    "samples2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>description</th>\n",
       "      <th>markers</th>\n",
       "      <th>platform</th>\n",
       "      <th>dataset_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2932488</th>\n",
       "      <td>B.Fem.Sp</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Female Splenic  B cells</td>\n",
       "      <td>CD19+Igm+TCRb-</td>\n",
       "      <td>Illumina NextSeq</td>\n",
       "      <td>Immgen-ULI-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932489</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Illumina NextSeq</td>\n",
       "      <td>Immgen-ULI-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932490</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Illumina NextSeq</td>\n",
       "      <td>Immgen-ULI-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932491</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Illumina NextSeq</td>\n",
       "      <td>Immgen-ULI-RNAseq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2932492</th>\n",
       "      <td>B.Fo.Sp</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Splenic Follicular</td>\n",
       "      <td>CD19+CD45R+IgM+CD93-CD23+CD43-CD5-</td>\n",
       "      <td>Illumina NextSeq</td>\n",
       "      <td>Immgen-ULI-RNAseq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      celltype    cell_lineage              description  \\\n",
       "sample_geo_accession                                                      \n",
       "GSM2932488            B.Fem.Sp  B Cell Lineage  Female Splenic  B cells   \n",
       "GSM2932489             B.Fo.Sp  B Cell Lineage       Splenic Follicular   \n",
       "GSM2932490             B.Fo.Sp  B Cell Lineage       Splenic Follicular   \n",
       "GSM2932491             B.Fo.Sp  B Cell Lineage       Splenic Follicular   \n",
       "GSM2932492             B.Fo.Sp  B Cell Lineage       Splenic Follicular   \n",
       "\n",
       "                                                 markers          platform  \\\n",
       "sample_geo_accession                                                         \n",
       "GSM2932488                                CD19+Igm+TCRb-  Illumina NextSeq   \n",
       "GSM2932489            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Illumina NextSeq   \n",
       "GSM2932490            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Illumina NextSeq   \n",
       "GSM2932491            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Illumina NextSeq   \n",
       "GSM2932492            CD19+CD45R+IgM+CD93-CD23+CD43-CD5-  Illumina NextSeq   \n",
       "\n",
       "                           dataset_name  \n",
       "sample_geo_accession                     \n",
       "GSM2932488            Immgen-ULI-RNAseq  \n",
       "GSM2932489            Immgen-ULI-RNAseq  \n",
       "GSM2932490            Immgen-ULI-RNAseq  \n",
       "GSM2932491            Immgen-ULI-RNAseq  \n",
       "GSM2932492            Immgen-ULI-RNAseq  "
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Immgen ULI RNAseq dataset - after standardisation \n",
    "samples2 = samples2[['celltype', 'cell_lineage', 'description', 'sort markers']]\n",
    "samples2.columns = ['celltype', 'cell_lineage', 'description', 'markers']\n",
    "samples2['platform'] = 'RNAseq (Illumina NextSeq)'\n",
    "samples2['dataset_name'] = 'Immgen-ULI-RNAseq'\n",
    "samples2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>tissue</th>\n",
       "      <th>description</th>\n",
       "      <th>surface_markers</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>iltp1786a-11</th>\n",
       "      <td>B2</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>Peritoneal</td>\n",
       "      <td>B2 B Cell</td>\n",
       "      <td>B220 Hi CD19+ F4/80-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-18</th>\n",
       "      <td>Mac</td>\n",
       "      <td>Macrophage Lineage</td>\n",
       "      <td>Peritoneal</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>Mac-1+ F4/80 Hi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-25</th>\n",
       "      <td>PreB</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>BM</td>\n",
       "      <td>B Cell Precursor</td>\n",
       "      <td>B220+ c-Kit- CD25+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-30</th>\n",
       "      <td>TN1</td>\n",
       "      <td>T Cell Lineage</td>\n",
       "      <td>Thymus</td>\n",
       "      <td>Triple Negative Thymocyte 1</td>\n",
       "      <td>Lin- c-Kit+ CD25- CD44+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-02</th>\n",
       "      <td>EffCD4T</td>\n",
       "      <td>T Cell Lineage</td>\n",
       "      <td>Spleen and LN</td>\n",
       "      <td>Effector Memory CD4 T Cells</td>\n",
       "      <td>CD4+ CD62L- CD25 Lo or - CD44+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             celltype        cell_lineage         tissue  \\\n",
       "sampleId                                                   \n",
       "iltp1786a-11       B2      B Cell Lineage     Peritoneal   \n",
       "iltp1786a-18      Mac  Macrophage Lineage     Peritoneal   \n",
       "iltp1786a-25     PreB      B Cell Lineage             BM   \n",
       "iltp1786a-30      TN1      T Cell Lineage         Thymus   \n",
       "iltp1786a-02  EffCD4T      T Cell Lineage  Spleen and LN   \n",
       "\n",
       "                              description                 surface_markers  \n",
       "sampleId                                                                   \n",
       "iltp1786a-11                    B2 B Cell            B220 Hi CD19+ F4/80-  \n",
       "iltp1786a-18                   Macrophage                 Mac-1+ F4/80 Hi  \n",
       "iltp1786a-25             B Cell Precursor              B220+ c-Kit- CD25+  \n",
       "iltp1786a-30  Triple Negative Thymocyte 1         Lin- c-Kit+ CD25- CD44+  \n",
       "iltp1786a-02  Effector Memory CD4 T Cells  CD4+ CD62L- CD25 Lo or - CD44+  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Haemopedia microarray dataset \n",
    "samples3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>description</th>\n",
       "      <th>markers</th>\n",
       "      <th>platform</th>\n",
       "      <th>dataset_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>iltp1786a-11</th>\n",
       "      <td>B2</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B2 B Cell</td>\n",
       "      <td>B220 Hi CD19+ F4/80-</td>\n",
       "      <td>Illumina WG6</td>\n",
       "      <td>Haemopedia-microarray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-18</th>\n",
       "      <td>Mac</td>\n",
       "      <td>Macrophage Lineage</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>Mac-1+ F4/80 Hi</td>\n",
       "      <td>Illumina WG6</td>\n",
       "      <td>Haemopedia-microarray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-25</th>\n",
       "      <td>PreB</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B Cell Precursor</td>\n",
       "      <td>B220+ c-Kit- CD25+</td>\n",
       "      <td>Illumina WG6</td>\n",
       "      <td>Haemopedia-microarray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-30</th>\n",
       "      <td>TN1</td>\n",
       "      <td>T Cell Lineage</td>\n",
       "      <td>Triple Negative Thymocyte 1</td>\n",
       "      <td>Lin- c-Kit+ CD25- CD44+</td>\n",
       "      <td>Illumina WG6</td>\n",
       "      <td>Haemopedia-microarray</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iltp1786a-02</th>\n",
       "      <td>EffCD4T</td>\n",
       "      <td>T Cell Lineage</td>\n",
       "      <td>Effector Memory CD4 T Cells</td>\n",
       "      <td>CD4+ CD62L- CD25 Lo or - CD44+</td>\n",
       "      <td>Illumina WG6</td>\n",
       "      <td>Haemopedia-microarray</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             celltype        cell_lineage                  description  \\\n",
       "sampleId                                                                 \n",
       "iltp1786a-11       B2      B Cell Lineage                    B2 B Cell   \n",
       "iltp1786a-18      Mac  Macrophage Lineage                   Macrophage   \n",
       "iltp1786a-25     PreB      B Cell Lineage             B Cell Precursor   \n",
       "iltp1786a-30      TN1      T Cell Lineage  Triple Negative Thymocyte 1   \n",
       "iltp1786a-02  EffCD4T      T Cell Lineage  Effector Memory CD4 T Cells   \n",
       "\n",
       "                                     markers      platform  \\\n",
       "sampleId                                                     \n",
       "iltp1786a-11            B220 Hi CD19+ F4/80-  Illumina WG6   \n",
       "iltp1786a-18                 Mac-1+ F4/80 Hi  Illumina WG6   \n",
       "iltp1786a-25              B220+ c-Kit- CD25+  Illumina WG6   \n",
       "iltp1786a-30         Lin- c-Kit+ CD25- CD44+  Illumina WG6   \n",
       "iltp1786a-02  CD4+ CD62L- CD25 Lo or - CD44+  Illumina WG6   \n",
       "\n",
       "                       dataset_name  \n",
       "sampleId                             \n",
       "iltp1786a-11  Haemopedia-microarray  \n",
       "iltp1786a-18  Haemopedia-microarray  \n",
       "iltp1786a-25  Haemopedia-microarray  \n",
       "iltp1786a-30  Haemopedia-microarray  \n",
       "iltp1786a-02  Haemopedia-microarray  "
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Haemopedia microarray dataset - after standardisation  \n",
    "samples3 = samples3[['celltype', 'cell_lineage', 'description', 'surface_markers']]\n",
    "samples3.columns = ['celltype', 'cell_lineage', 'description', 'markers']\n",
    "samples3['platform'] = 'Microarray (Illumina WG6)'\n",
    "samples3['dataset_name'] = 'Haemopedia-microarray'\n",
    "samples3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>description</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>surface_markers</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM854035</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854036</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854037</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854038</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854039</th>\n",
       "      <td>MPPa</td>\n",
       "      <td>Multi-potent Progenitor subset A</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>CD34+ Slamf1+ Tie2+ Vcam1+ Sca-1+ c-Kit+ Il7ra...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          celltype                       description  \\\n",
       "sampleId                                               \n",
       "GSM854035      HSC          Hematopoietic Stem Cells   \n",
       "GSM854036      HSC          Hematopoietic Stem Cells   \n",
       "GSM854037      HSC          Hematopoietic Stem Cells   \n",
       "GSM854038      HSC          Hematopoietic Stem Cells   \n",
       "GSM854039     MPPa  Multi-potent Progenitor subset A   \n",
       "\n",
       "                         cell_lineage  \\\n",
       "sampleId                                \n",
       "GSM854035  Multi Potential Progenitor   \n",
       "GSM854036  Multi Potential Progenitor   \n",
       "GSM854037  Multi Potential Progenitor   \n",
       "GSM854038  Multi Potential Progenitor   \n",
       "GSM854039  Multi Potential Progenitor   \n",
       "\n",
       "                                             surface_markers  \n",
       "sampleId                                                      \n",
       "GSM854035  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...  \n",
       "GSM854036  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...  \n",
       "GSM854037  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...  \n",
       "GSM854038  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...  \n",
       "GSM854039  CD34+ Slamf1+ Tie2+ Vcam1+ Sca-1+ c-Kit+ Il7ra...  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Gene-Expression-Commons dataset \n",
    "samples4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>description</th>\n",
       "      <th>markers</th>\n",
       "      <th>platform</th>\n",
       "      <th>dataset_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM854035</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "      <td>Affymetrix Mouse430_2</td>\n",
       "      <td>Gene-Expression-Commons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854036</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "      <td>Affymetrix Mouse430_2</td>\n",
       "      <td>Gene-Expression-Commons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854037</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "      <td>Affymetrix Mouse430_2</td>\n",
       "      <td>Gene-Expression-Commons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854038</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>Hematopoietic Stem Cells</td>\n",
       "      <td>CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...</td>\n",
       "      <td>Affymetrix Mouse430_2</td>\n",
       "      <td>Gene-Expression-Commons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM854039</th>\n",
       "      <td>MPPa</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>Multi-potent Progenitor subset A</td>\n",
       "      <td>CD34+ Slamf1+ Tie2+ Vcam1+ Sca-1+ c-Kit+ Il7ra...</td>\n",
       "      <td>Affymetrix Mouse430_2</td>\n",
       "      <td>Gene-Expression-Commons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          celltype                cell_lineage  \\\n",
       "sampleId                                         \n",
       "GSM854035      HSC  Multi Potential Progenitor   \n",
       "GSM854036      HSC  Multi Potential Progenitor   \n",
       "GSM854037      HSC  Multi Potential Progenitor   \n",
       "GSM854038      HSC  Multi Potential Progenitor   \n",
       "GSM854039     MPPa  Multi Potential Progenitor   \n",
       "\n",
       "                                description  \\\n",
       "sampleId                                      \n",
       "GSM854035          Hematopoietic Stem Cells   \n",
       "GSM854036          Hematopoietic Stem Cells   \n",
       "GSM854037          Hematopoietic Stem Cells   \n",
       "GSM854038          Hematopoietic Stem Cells   \n",
       "GSM854039  Multi-potent Progenitor subset A   \n",
       "\n",
       "                                                     markers  \\\n",
       "sampleId                                                       \n",
       "GSM854035  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...   \n",
       "GSM854036  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...   \n",
       "GSM854037  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...   \n",
       "GSM854038  CD34- Flk2- Sca-1+ c-Kit+ Il7ra- Lin(CD4, CD8,...   \n",
       "GSM854039  CD34+ Slamf1+ Tie2+ Vcam1+ Sca-1+ c-Kit+ Il7ra...   \n",
       "\n",
       "                        platform             dataset_name  \n",
       "sampleId                                                   \n",
       "GSM854035  Affymetrix Mouse430_2  Gene-Expression-Commons  \n",
       "GSM854036  Affymetrix Mouse430_2  Gene-Expression-Commons  \n",
       "GSM854037  Affymetrix Mouse430_2  Gene-Expression-Commons  \n",
       "GSM854038  Affymetrix Mouse430_2  Gene-Expression-Commons  \n",
       "GSM854039  Affymetrix Mouse430_2  Gene-Expression-Commons  "
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Gene-Expression-Commons dataset - after standardisation \n",
    "samples4 = samples4[['celltype', 'cell_lineage', 'description', 'surface_markers']]\n",
    "samples4.columns = ['celltype', 'cell_lineage', 'description', 'markers']\n",
    "samples4['platform'] = 'Microarray (Affymetrix Mouse430_2)'\n",
    "samples4['dataset_name'] = 'Gene-Expression-Commons'\n",
    "samples4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>immGenLineage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B1a_PC</th>\n",
       "      <td>B1a_PC</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B1a_Sp</th>\n",
       "      <td>B1a_Sp</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B1b_PC</th>\n",
       "      <td>B1b_PC</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BEC_MLN</th>\n",
       "      <td>BEC_MLN</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BEC_SLN</th>\n",
       "      <td>BEC_SLN</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         celltype    cell_lineage immGenLineage\n",
       "sampleId                                       \n",
       "B1a_PC     B1a_PC  B Cell Lineage       B cells\n",
       "B1a_Sp     B1a_Sp  B Cell Lineage       B cells\n",
       "B1b_PC     B1b_PC  B Cell Lineage       B cells\n",
       "BEC_MLN   BEC_MLN  B Cell Lineage       B cells\n",
       "BEC_SLN   BEC_SLN  B Cell Lineage       B cells"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Immgen microarray dataset\n",
    "samples5.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# metadata of Gene-Expression-Commons dataset - after standardisation \n",
    "samples5 = samples5[['celltype', 'cell_lineage']]\n",
    "samples5['platform'] = 'Microarray (Affymetrix Mouse Gene 1.0 ST Array)'\n",
    "samples5['dataset_name'] = 'Immgen-microarray'\n",
    "samples5.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>surface_markers</th>\n",
       "      <th>tissue</th>\n",
       "      <th>description</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-Cell(1)</th>\n",
       "      <td>B-Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>CD19+ 33D1-</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>B-Cell</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-Cell(2)</th>\n",
       "      <td>B-Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>CD19+ 33D1-</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>B-Cell</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Granulocyte(1)</th>\n",
       "      <td>Granulocyte</td>\n",
       "      <td>Neutrophil Lineage</td>\n",
       "      <td>Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...</td>\n",
       "      <td>BM</td>\n",
       "      <td>Granulocytes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Granulocyte(2)</th>\n",
       "      <td>Granulocyte</td>\n",
       "      <td>Neutrophil Lineage</td>\n",
       "      <td>Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...</td>\n",
       "      <td>BM</td>\n",
       "      <td>Granulocytes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSC(1)</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>SP+ c-Kit+ Sca-1+ Lin-</td>\n",
       "      <td>BM</td>\n",
       "      <td>LT-HSC</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   celltype                cell_lineage  \\\n",
       "sampleId                                                  \n",
       "B-Cell(1)            B-Cell              B Cell Lineage   \n",
       "B-Cell(2)            B-Cell              B Cell Lineage   \n",
       "Granulocyte(1)  Granulocyte          Neutrophil Lineage   \n",
       "Granulocyte(2)  Granulocyte          Neutrophil Lineage   \n",
       "HSC(1)                  HSC  Multi Potential Progenitor   \n",
       "\n",
       "                                                  surface_markers  tissue  \\\n",
       "sampleId                                                                    \n",
       "B-Cell(1)                                             CD19+ 33D1-  Spleen   \n",
       "B-Cell(2)                                             CD19+ 33D1-  Spleen   \n",
       "Granulocyte(1)  Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...      BM   \n",
       "Granulocyte(2)  Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...      BM   \n",
       "HSC(1)                                     SP+ c-Kit+ Sca-1+ Lin-      BM   \n",
       "\n",
       "                 description notes  \n",
       "sampleId                            \n",
       "B-Cell(1)             B-Cell   NaN  \n",
       "B-Cell(2)             B-Cell   NaN  \n",
       "Granulocyte(1)  Granulocytes   NaN  \n",
       "Granulocyte(2)  Granulocytes   NaN  \n",
       "HSC(1)                LT-HSC   NaN  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Goodell dataset\n",
    "samples6.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/monica/anaconda3/envs/stemformatics/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>description</th>\n",
       "      <th>markers</th>\n",
       "      <th>platform</th>\n",
       "      <th>dataset_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-Cell(1)</th>\n",
       "      <td>B-Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B-Cell</td>\n",
       "      <td>CD19+ 33D1-</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>Goodell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B-Cell(2)</th>\n",
       "      <td>B-Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>B-Cell</td>\n",
       "      <td>CD19+ 33D1-</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>Goodell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Granulocyte(1)</th>\n",
       "      <td>Granulocyte</td>\n",
       "      <td>Neutrophil Lineage</td>\n",
       "      <td>Granulocytes</td>\n",
       "      <td>Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>Goodell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Granulocyte(2)</th>\n",
       "      <td>Granulocyte</td>\n",
       "      <td>Neutrophil Lineage</td>\n",
       "      <td>Granulocytes</td>\n",
       "      <td>Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>Goodell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSC(1)</th>\n",
       "      <td>HSC</td>\n",
       "      <td>Multi Potential Progenitor</td>\n",
       "      <td>LT-HSC</td>\n",
       "      <td>SP+ c-Kit+ Sca-1+ Lin-</td>\n",
       "      <td>Microarray (Affymetrix Mouse430_2)</td>\n",
       "      <td>Goodell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   celltype                cell_lineage   description  \\\n",
       "sampleId                                                                \n",
       "B-Cell(1)            B-Cell              B Cell Lineage        B-Cell   \n",
       "B-Cell(2)            B-Cell              B Cell Lineage        B-Cell   \n",
       "Granulocyte(1)  Granulocyte          Neutrophil Lineage  Granulocytes   \n",
       "Granulocyte(2)  Granulocyte          Neutrophil Lineage  Granulocytes   \n",
       "HSC(1)                  HSC  Multi Potential Progenitor        LT-HSC   \n",
       "\n",
       "                                                          markers  \\\n",
       "sampleId                                                            \n",
       "B-Cell(1)                                             CD19+ 33D1-   \n",
       "B-Cell(2)                                             CD19+ 33D1-   \n",
       "Granulocyte(1)  Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...   \n",
       "Granulocyte(2)  Gr-1+ 7/4 clone+ Lin- (CD2,CD5,B220,F4/80,ICAM...   \n",
       "HSC(1)                                     SP+ c-Kit+ Sca-1+ Lin-   \n",
       "\n",
       "                                          platform dataset_name  \n",
       "sampleId                                                         \n",
       "B-Cell(1)       Microarray (Affymetrix Mouse430_2)      Goodell  \n",
       "B-Cell(2)       Microarray (Affymetrix Mouse430_2)      Goodell  \n",
       "Granulocyte(1)  Microarray (Affymetrix Mouse430_2)      Goodell  \n",
       "Granulocyte(2)  Microarray (Affymetrix Mouse430_2)      Goodell  \n",
       "HSC(1)          Microarray (Affymetrix Mouse430_2)      Goodell  "
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata of Goodell dataset - after standardisation \n",
    "samples6 = samples6[['celltype', 'cell_lineage', 'description', 'surface_markers']]\n",
    "samples6.columns = ['celltype', 'cell_lineage', 'description', 'markers']\n",
    "samples6['platform'] = 'Microarray (Affymetrix Mouse430_2)'\n",
    "samples6['dataset_name'] = 'Goodell'\n",
    "samples6.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
